File: llm-context/run_evaluation.py
"""
Evaluation Runner for Covenant LLM Generation

Runs the full test suite and generates analysis reports.
"""

import argparse
import json
from pathlib import Path
from datetime import datetime

from test_suite import create_test_suite, get_test_suite_sample
from generation_harness import (
    run_test_suite, print_summary, GenerationMetrics, ModelProvider
)


def analyze_results(results_file: Path):
    """
    Analyze results from a test run and generate detailed report.

    Args:
        results_file: Path to JSONL results file
    """
    results = []
    with open(results_file, 'r', encoding='utf-8') as f:
        for line in f:
            if line.strip():
                results.append(json.loads(line))

    total = len(results)
    if total == 0:
        print("No results to analyze")
        return

    # Overall metrics
    first_pass = sum(1 for r in results if r['first_pass_success'])
    final = sum(1 for r in results if r['final_success'])

    total_prompt = sum(r['total_prompt_tokens'] for r in results)
    total_completion = sum(r['total_completion_tokens'] for r in results)
    total_cost = sum(r['total_cost_usd'] for r in results)
    total_time = sum(r['total_duration_ms'] for r in results)

    avg_attempts = sum(r['total_attempts'] for r in results) / total
    avg_cost = total_cost / total
    avg_time = total_time / total

    # Print report
    print("\n" + "="*80)
    print("EVALUATION REPORT")
    print("="*80)
    print(f"Date: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}")
    print(f"Results file: {results_file}")
    print()

    print("OVERALL METRICS")
    print("-"*80)
    print(f"Total tasks: {total}")
    print(f"First-pass success: {first_pass}/{total} ({first_pass/total*100:.1f}%)")
    print(f"Final success: {final}/{total} ({final/total*100:.1f}%)")
    print(f"Improvement: {final - first_pass} tasks (+{(final-first_pass)/total*100:.1f}%)")
    print()

    print("RESOURCE USAGE")
    print("-"*80)
    print(f"Total prompt tokens: {total_prompt:,}")
    print(f"Total completion tokens: {total_completion:,}")
    print(f"Total tokens: {total_prompt + total_completion:,}")
    print(f"Total cost: ${total_cost:.2f}")
    print(f"Total time: {total_time/1000:.1f}s")
    print()

    print("AVERAGES PER TASK")
    print("-"*80)
    print(f"Average attempts: {avg_attempts:.2f}")
    print(f"Average cost: ${avg_cost:.4f}")
    print(f"Average time: {avg_time:.0f}ms")
    print(f"Average tokens: {(total_prompt + total_completion)/total:.0f}")
    print()

    # Success rate by task type
    print("SUCCESS RATE BY TASK TYPE")
    print("-"*80)
    from collections import defaultdict
    by_type = defaultdict(lambda: {'total': 0, 'first': 0, 'final': 0})

    for r in results:
        task_type = r['task_type']
        by_type[task_type]['total'] += 1
        if r['first_pass_success']:
            by_type[task_type]['first'] += 1
        if r['final_success']:
            by_type[task_type]['final'] += 1

    for task_type in sorted(by_type.keys()):
        stats = by_type[task_type]
        total_t = stats['total']
        first_t = stats['first']
        final_t = stats['final']
        print(f"{task_type:20s}: {final_t:3d}/{total_t:3d} ({final_t/total_t*100:5.1f}%) "
              f"[first: {first_t/total_t*100:5.1f}%]")
    print()

    # Error analysis
    print("ERROR ANALYSIS")
    print("-"*80)
    from collections import Counter
    all_errors = []
    for r in results:
        all_errors.extend(r.get('error_codes', []))

    if all_errors:
        print("Most common errors:")
        for code, count in Counter(all_errors).most_common(15):
            print(f"  {code:20s}: {count:3d} ({count/len([r for r in results if not r['final_success']])*100:.1f}% of failures)")
    else:
        print("No errors (all tasks succeeded!)")
    print()

    # Cost breakdown
    print("COST BREAKDOWN")
    print("-"*80)

    # By complexity (measured by attempts needed)
    simple = [r for r in results if r['total_attempts'] == 1]
    medium = [r for r in results if 1 < r['total_attempts'] <= 2]
    complex_tasks = [r for r in results if r['total_attempts'] > 2]

    print(f"Simple (1 attempt):  {len(simple):3d} tasks, ${sum(r['total_cost_usd'] for r in simple):.2f} total, "
          f"${sum(r['total_cost_usd'] for r in simple)/len(simple):.4f} avg" if simple else "")
    print(f"Medium (2 attempts): {len(medium):3d} tasks, ${sum(r['total_cost_usd'] for r in medium):.2f} total, "
          f"${sum(r['total_cost_usd'] for r in medium)/len(medium):.4f} avg" if medium else "")
    print(f"Complex (3+ attempts): {len(complex_tasks):3d} tasks, ${sum(r['total_cost_usd'] for r in complex_tasks):.2f} total, "
          f"${sum(r['total_cost_usd'] for r in complex_tasks)/len(complex_tasks):.4f} avg" if complex_tasks else "")
    print()

    # Failure examples
    failures = [r for r in results if not r['final_success']]
    if failures:
        print(f"FAILURE EXAMPLES (showing first 5 of {len(failures)})")
        print("-"*80)
        for r in failures[:5]:
            print(f"{r['task_id']}:")
            print(f"  Type: {r['task_type']}")
            print(f"  Attempts: {r['total_attempts']}")
            print(f"  Errors: {', '.join(r.get('error_codes', ['unknown']))}")
            print()

    # Export summary
    summary_file = results_file.with_suffix('.summary.json')
    summary = {
        'date': datetime.now().isoformat(),
        'total_tasks': total,
        'first_pass_success': first_pass,
        'first_pass_rate': first_pass / total,
        'final_success': final,
        'final_success_rate': final / total,
        'avg_attempts': avg_attempts,
        'avg_cost_usd': avg_cost,
        'avg_time_ms': avg_time,
        'total_cost_usd': total_cost,
        'total_time_ms': total_time,
        'by_type': {
            task_type: {
                'total': stats['total'],
                'first_pass_success': stats['first'],
                'final_success': stats['final'],
                'first_pass_rate': stats['first'] / stats['total'],
                'final_success_rate': stats['final'] / stats['total'],
            }
            for task_type, stats in by_type.items()
        },
        'top_errors': Counter(all_errors).most_common(10),
    }

    with open(summary_file, 'w', encoding='utf-8') as f:
        json.dump(summary, f, indent=2)

    print(f"Summary exported to: {summary_file}")
    print("="*80)


def main():
    parser = argparse.ArgumentParser(description="Run Covenant LLM generation evaluation")
    parser.add_argument('--sample', type=int, help="Run on random sample of N tasks")
    parser.add_argument('--category', type=str, help="Run only tasks of specific category")
    parser.add_argument('--output', type=str, help="Output file for results (JSONL)")
    parser.add_argument('--analyze', type=str, help="Analyze existing results file")
    parser.add_argument('--verbose', '-v', action='store_true', help="Verbose output")
    parser.add_argument('--provider', type=str, default='mock',
                       choices=['anthropic', 'openai', 'mock'],
                       help="LLM provider (default: mock)")

    args = parser.parse_args()

    # Analysis mode
    if args.analyze:
        analyze_results(Path(args.analyze))
        return

    # Generation mode
    print("Covenant LLM Generation Evaluation")
    print("="*80)

    # Select tasks
    if args.sample:
        print(f"Running sample of {args.sample} tasks")
        tasks = get_test_suite_sample(args.sample)
    elif args.category:
        from example_selector import TaskType
        task_type = TaskType(args.category)
        from test_suite import get_test_suite_by_category
        tasks = get_test_suite_by_category(task_type)
        print(f"Running {len(tasks)} tasks from category: {args.category}")
    else:
        tasks = create_test_suite()
        print(f"Running full test suite: {len(tasks)} tasks")

    # Set output file
    if args.output:
        output_file = args.output
    else:
        timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
        output_file = f"results_{timestamp}.jsonl"

    print(f"Provider: {args.provider}")
    print(f"Output: {output_file}")
    print()

    if args.provider != 'mock':
        print("WARNING: This will make real API calls and incur costs!")
        print(f"Estimated cost: ${len(tasks) * 0.15:.2f} - ${len(tasks) * 0.30:.2f}")
        response = input("Continue? (yes/no): ")
        if response.lower() not in ['yes', 'y']:
            print("Aborted")
            return

    # Run evaluation
    print("\nStarting evaluation...")
    print("="*80)

    results = run_test_suite(tasks, output_file=output_file, verbose=args.verbose)

    # Print summary
    print_summary(results)

    # Generate detailed analysis
    print("\nGenerating detailed analysis...")
    analyze_results(Path(output_file))


if __name__ == "__main__":
    main()

File: tools/llm-eval/src/run-evaluation.ts
/**
 * Evaluation Runner for Covenant LLM Generation.
 *
 * CLI entry point that runs the full test suite and generates analysis reports.
 */

import { parseArgs } from "jsr:@std/cli@1/parse-args";
import type { EvaluationOptions, GenerationMetrics, ModelProvider, TaskType } from "./types.ts";
import { createTestSuite, getTestSuiteByCategory, getTestSuiteSample } from "./test-suite.ts";
import { GenerationHarness, printSummary, runTestSuite } from "./generation-harness.ts";
import { ExampleSelector } from "./example-selector.ts";
import { CompilerValidator } from "./compiler-validator.ts";

/** Load and analyze results from a JSONL file. */
async function analyzeResults(resultsPath: string): Promise<void> {
  const text = await Deno.readTextFile(resultsPath);
  const results: GenerationMetrics[] = text
    .split("\n")
    .filter((line) => line.trim())
    .map((line) => JSON.parse(line));

  const total = results.length;
  if (total === 0) {
    console.log("No results to analyze");
    return;
  }

  const firstPass = results.filter((r) => r.firstPassSuccess).length;
  const final = results.filter((r) => r.finalSuccess).length;
  const totalPrompt = results.reduce((s, r) => s + r.totalPromptTokens, 0);
  const totalCompletion = results.reduce((s, r) => s + r.totalCompletionTokens, 0);
  const totalCost = results.reduce((s, r) => s + r.totalCostUsd, 0);
  const totalTime = results.reduce((s, r) => s + r.totalDurationMs, 0);
  const avgAttempts = results.reduce((s, r) => s + r.totalAttempts, 0) / total;
  const avgCost = totalCost / total;
  const avgTime = totalTime / total;

  console.log(`\n${"=".repeat(80)}`);
  console.log("EVALUATION REPORT");
  console.log("=".repeat(80));
  console.log(`Date: ${new Date().toISOString()}`);
  console.log(`Results file: ${resultsPath}`);
  console.log();

  console.log("OVERALL METRICS");
  console.log("-".repeat(80));
  console.log(`Total tasks: ${total}`);
  console.log(`First-pass success: ${firstPass}/${total} (${(firstPass / total * 100).toFixed(1)}%)`);
  console.log(`Final success: ${final}/${total} (${(final / total * 100).toFixed(1)}%)`);
  console.log(`Improvement: ${final - firstPass} tasks (+${((final - firstPass) / total * 100).toFixed(1)}%)`);
  console.log();

  console.log("RESOURCE USAGE");
  console.log("-".repeat(80));
  console.log(`Total prompt tokens: ${totalPrompt.toLocaleString()}`);
  console.log(`Total completion tokens: ${totalCompletion.toLocaleString()}`);
  console.log(`Total tokens: ${(totalPrompt + totalCompletion).toLocaleString()}`);
  console.log(`Total cost: $${totalCost.toFixed(2)}`);
  console.log(`Total time: ${(totalTime / 1000).toFixed(1)}s`);
  console.log();

  console.log("AVERAGES PER TASK");
  console.log("-".repeat(80));
  console.log(`Average attempts: ${avgAttempts.toFixed(2)}`);
  console.log(`Average cost: $${avgCost.toFixed(4)}`);
  console.log(`Average time: ${avgTime.toFixed(0)}ms`);
  console.log(`Average tokens: ${((totalPrompt + totalCompletion) / total).toFixed(0)}`);
  console.log();

  // Success rate by task type
  console.log("SUCCESS RATE BY TASK TYPE");
  console.log("-".repeat(80));
  const byType = new Map<string, { total: number; first: number; final: number }>();
  for (const r of results) {
    const entry = byType.get(r.taskType) ?? { total: 0, first: 0, final: 0 };
    entry.total++;
    if (r.firstPassSuccess) entry.first++;
    if (r.finalSuccess) entry.final++;
    byType.set(r.taskType, entry);
  }

  for (const [taskType, stats] of [...byType.entries()].sort()) {
    const finalPct = (stats.final / stats.total * 100).toFixed(1);
    const firstPct = (stats.first / stats.total * 100).toFixed(1);
    console.log(
      `${taskType.padEnd(20)}: ${String(stats.final).padStart(3)}/${String(stats.total).padStart(3)} (${finalPct.padStart(5)}%) [first: ${firstPct.padStart(5)}%]`,
    );
  }
  console.log();

  // Error analysis
  console.log("ERROR ANALYSIS");
  console.log("-".repeat(80));
  const errorCounts = new Map<string, number>();
  for (const r of results) {
    for (const code of r.errorCodes) {
      errorCounts.set(code, (errorCounts.get(code) ?? 0) + 1);
    }
  }

  if (errorCounts.size > 0) {
    console.log("Most common errors:");
    const sorted = [...errorCounts.entries()].sort((a, b) => b[1] - a[1]);
    for (const [code, count] of sorted.slice(0, 15)) {
      console.log(`  ${code.padEnd(20)}: ${count}`);
    }
  } else {
    console.log("No errors (all tasks succeeded!)");
  }
  console.log();

  // Cost breakdown
  console.log("COST BREAKDOWN");
  console.log("-".repeat(80));
  const simple = results.filter((r) => r.totalAttempts === 1);
  const medium = results.filter((r) => r.totalAttempts === 2);
  const complex = results.filter((r) => r.totalAttempts > 2);

  if (simple.length > 0) {
    const cost = simple.reduce((s, r) => s + r.totalCostUsd, 0);
    console.log(`Simple (1 attempt):    ${String(simple.length).padStart(3)} tasks, $${cost.toFixed(2)} total, $${(cost / simple.length).toFixed(4)} avg`);
  }
  if (medium.length > 0) {
    const cost = medium.reduce((s, r) => s + r.totalCostUsd, 0);
    console.log(`Medium (2 attempts):   ${String(medium.length).padStart(3)} tasks, $${cost.toFixed(2)} total, $${(cost / medium.length).toFixed(4)} avg`);
  }
  if (complex.length > 0) {
    const cost = complex.reduce((s, r) => s + r.totalCostUsd, 0);
    console.log(`Complex (3+ attempts): ${String(complex.length).padStart(3)} tasks, $${cost.toFixed(2)} total, $${(cost / complex.length).toFixed(4)} avg`);
  }
  console.log();

  // Failure examples
  const failures = results.filter((r) => !r.finalSuccess);
  if (failures.length > 0) {
    console.log(`FAILURE EXAMPLES (showing first 5 of ${failures.length})`);
    console.log("-".repeat(80));
    for (const r of failures.slice(0, 5)) {
      console.log(`${r.taskId}:`);
      console.log(`  Type: ${r.taskType}`);
      console.log(`  Attempts: ${r.totalAttempts}`);
      console.log(`  Errors: ${r.errorCodes.join(", ") || "unknown"}`);
      console.log();
    }
  }

  // Export summary JSON
  const summaryPath = resultsPath.replace(/\.jsonl$/, ".summary.json");
  const summary = {
    date: new Date().toISOString(),
    totalTasks: total,
    firstPassSuccess: firstPass,
    firstPassRate: firstPass / total,
    finalSuccess: final,
    finalSuccessRate: final / total,
    avgAttempts,
    avgCostUsd: avgCost,
    avgTimeMs: avgTime,
    totalCostUsd: totalCost,
    totalTimeMs: totalTime,
    byType: Object.fromEntries(
      [...byType.entries()].map(([k, v]) => [k, {
        total: v.total,
        firstPassSuccess: v.first,
        finalSuccess: v.final,
        firstPassRate: v.first / v.total,
        finalSuccessRate: v.final / v.total,
      }]),
    ),
    topErrors: [...errorCounts.entries()].sort((a, b) => b[1] - a[1]).slice(0, 10),
  };

  await Deno.writeTextFile(summaryPath, JSON.stringify(summary, null, 2));
  console.log(`Summary exported to: ${summaryPath}`);
  console.log("=".repeat(80));
}

function printUsage(): void {
  console.log(`Usage: deno task eval [options]

Options:
  --provider, -p   Model provider: anthropic | openai | mock (default: mock)
  --model, -m      Model ID (default: claude-sonnet-4-5-20250929)
  --sample, -n     Run N random tasks
  --category       Run only tasks of this category
  --output, -o     JSONL output file path
  --analyze        Analyze existing results file (no generation)
  --verbose, -v    Verbose output
  --help, -h       Show help
`);
}

async function main(): Promise<void> {
  const args = parseArgs(Deno.args, {
    string: ["provider", "model", "category", "output", "analyze", "sample"],
    boolean: ["verbose", "help"],
    alias: {
      p: "provider",
      m: "model",
      n: "sample",
      o: "output",
      v: "verbose",
      h: "help",
    },
    default: {
      provider: "mock",
      model: "claude-sonnet-4-5-20250929",
    },
  });

  if (args.help) {
    printUsage();
    Deno.exit(0);
  }

  // Analysis mode
  if (args.analyze) {
    await analyzeResults(args.analyze);
    return;
  }

  // Generation mode
  console.log("Covenant LLM Generation Evaluation");
  console.log("=".repeat(80));

  // Select tasks
  let tasks;
  if (args.sample) {
    const n = parseInt(args.sample, 10);
    console.log(`Running sample of ${n} tasks`);
    tasks = getTestSuiteSample(n);
  } else if (args.category) {
    tasks = getTestSuiteByCategory(args.category as TaskType);
    console.log(`Running ${tasks.length} tasks from category: ${args.category}`);
  } else {
    tasks = createTestSuite();
    console.log(`Running full test suite: ${tasks.length} tasks`);
  }

  // Output file
  const outputFile = args.output ?? `results_${new Date().toISOString().replace(/[:.]/g, "").slice(0, 15)}.jsonl`;

  const provider = args.provider as ModelProvider;
  console.log(`Provider: ${provider}`);
  console.log(`Model: ${args.model}`);
  console.log(`Output: ${outputFile}`);
  console.log();

  // Cost warning for real providers
  if (provider !== "mock") {
    console.log("WARNING: This will make real API calls and incur costs!");
    console.log(`Estimated cost: $${(tasks.length * 0.15).toFixed(2)} - $${(tasks.length * 0.30).toFixed(2)}`);

    const buf = new Uint8Array(10);
    await Deno.stdout.write(new TextEncoder().encode("Continue? (yes/no): "));
    const n = await Deno.stdin.read(buf);
    const response = new TextDecoder().decode(buf.subarray(0, n ?? 0)).trim().toLowerCase();
    if (response !== "yes" && response !== "y") {
      console.log("Aborted");
      return;
    }
  }

  // Load spec
  const specPath = new URL("../../llm-context/SPEC_CONDENSED.md", import.meta.url).pathname;
  let specContent: string;
  try {
    specContent = await Deno.readTextFile(specPath);
  } catch {
    console.warn(`Spec file not found at ${specPath}, using empty spec`);
    specContent = "";
  }

  // Build options
  const options: EvaluationOptions = {
    model: { provider, modelId: args.model!, maxOutputTokens: 4000 },
    maxCorrectionRounds: 2,
    maxExampleTokens: 1500,
    verbose: args.verbose ?? false,
    outputFile,
    sample: args.sample ? parseInt(args.sample, 10) : undefined,
    category: args.category as TaskType | undefined,
  };

  // Initialize components
  const validator = new CompilerValidator();
  const exampleSelector = new ExampleSelector();
  const harness = new GenerationHarness(
    options.model,
    validator,
    exampleSelector,
    specContent,
    options.maxCorrectionRounds,
  );

  // Run
  console.log("\nStarting evaluation...");
  console.log("=".repeat(80));

  const results = await runTestSuite(tasks, harness, outputFile, options.verbose);

  // Summary
  printSummary(results);

  // Detailed analysis
  console.log("\nGenerating detailed analysis...");
  await analyzeResults(outputFile);
}

if (import.meta.main) {
  main();
}
